crates/wysiwyg/src/composer_model/hyperlinks.rs

```diff
@@ -58,9 +61,63 @@ where
 
     pub fn set_link(&mut self, link: S) -> ComposerUpdate<S> {
         self.push_state_to_history();
-        let (s, e) = self.safe_selection();
-        let range = self.state.dom.find_range(s, e);
-        self.set_link_range(range, link)
+        let (mut s, mut e) = self.safe_selection();
+
+        let mut range = self.state.dom.find_range(s, e);
+
+        // Find container link that completely covers the range
+        if let Some(link) = self.find_parent_links(&range) {
+            // If found, update the range to the container link bounds
+            range = self.state.dom.find_range_by_node(&link);
+            (s, e) = (range.start(), range.end());
+        }
+
+        if s == e {
+            return ComposerUpdate::keep();
+        }
+
+        let inserted = self
+            .state
+            .dom
+            .insert_parent(&range, DomNode::new_link(link, vec![]));
+
+        // Ensure no duplication by deleting any links contained within the new link
+        self.delete_child_links(&inserted);
+
+        self.create_update_replace_all()
+    }
+
+    fn delete_child_links(&mut self, node_handle: &DomHandle) {
```

Comment:
I wasn't sure if this method belonged in the composer layer or whether it might be better abstracted into the DOM layer. Open to ideas!

Reply:
If we could pass the `ContainerNodeKind` to match it could make sense to make it `pub(crate)` and add it to the Dom layer instead as `remove_child_nodes_of_kind(...)` or something like that.

Reply:
I tried to refactor this into something a bit more general but it got a bit complicated when we start needing to handle non container nodes (and is too specific IMO to warrant refactoring if we handle just the containers).

Reply:
I gave it a try too and It's way more complex than I thought, so no worries, this is fine.